Refactoring cli related code

levkohimins opened the pull request:
<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Refactoring CLI related code with a CLI package implementation similar to `github.com/urfave/cli`, but with the main difference, it does not return an error if the flag or command is not found, instead returns them as argument values. This also affected the following:
1. Structuring flags and commands
1. Automatic CLI Help Generation
1. Splitted `RunTerragrunt` code by commands
1. Get rid of repetitive code
1. Better code readability


Fixes #2599

<!-- Description of the changes introduced by this PR. -->

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

<!-- One-line description of the PR that can be included in the final release notes. -->
Added / Removed / Updated [X].

brikis98 commented on cli/commands/run-all/command.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+func newSubCommands(opts *options.TerragruntOptions) cli.Commands {
+	cmds := cli.Commands{
+		terragruntinfo.NewCommand(opts),    // terragrunt-info
+		validateinputs.NewCommand(opts),    // validate-inputs
+		graphdependencies.NewCommand(opts), // graph-dependencies
+		hclfmt.NewCommand(opts),            // hclfmt
+		renderjson.NewCommand(opts),        // render-json
+		awsproviderpatch.NewCommand(opts),  // aws-provider-patch
+	}
+
+	sort.Sort(cmds)
+
+	// add terraform command `*` after sorting to put the command at the end of the list in the help.
+	cmds.Add(terraform.NewCommand(opts))
```
This logic looks very similar to what's in `app.go`... Should it be shared somehow?

levkohimins replied:
We need add one more command after sorting and there is no `run-all` command in this list as in `app.go`
Unfortunately, any attempt to share this logic only worsens the readability of the code or creates a cycle.

brikis98 commented on cli/commands/terraform/target.go:
```diff
@@ -0,0 +1,39 @@
+
+type TargetCallbackType func(opts *options.TerragruntOptions, config *config.TerragruntConfig) error
+
+// Since most terragrunt commands like `render-json`, `aws-provider-patch` ...  require preparatory steps.
+// Target helps to execute a command up to a certain point and then run a callback function.
```
I'm not sure I understand this or what `Target` represents. Same for `isPoint` or the need for callbacks. 

Can you provide a bit more context behind this structure?

brikis98 replied:
Got it, thanks for the explanation. I agree that `runTerraform` is way too big and complicated and needs to be simplified. I'm a bit worried that the target/isPoint approach actually makes it _more_ complicated. But if it's an incremental stepping stone—a way to avoid making this PR even bigger and take even longer—I'm OK with that, as long as the next step is to refactor it into something simpler / more modular, and remove the target/isPoint behavior.

levkohimins replied:
Totally agree with everything you said. The next step will be refactoring the `runTerraform` function.

brikis98 commented on docs/_docs/02_features/auto-init.md:
```diff
@@ -27,6 +27,6 @@ As [mentioned]({{site.baseurl}}/docs/features/keep-your-cli-flags-dry/#extra_arg
 Note that there might be cases where terragrunt does not properly detect that `terraform init` needs be called. In this case, terraform would fail. Running `terragrunt init` again corrects this situation.
 
-For some use cases, it might be desirable to disable Auto-Init. For example, if each user wants to specify a different `-plugin-dir` option to `terraform init` (and therefore it cannot be put in `extra_arguments`). To disable Auto-Init, use the `--terragrunt-no-auto-init` command line option or set the `TERRAGRUNT_AUTO_INIT` environment variable to `false`.
+For some use cases, it might be desirable to disable Auto-Init. For example, if each user wants to specify a different `-plugin-dir` option to `terraform init` (and therefore it cannot be put in `extra_arguments`). To disable Auto-Init, use the `--terragrunt-no-auto-init` command line option or set the `TERRAGRUNT_NO_AUTO_INIT` environment variable to `true`.
```
Did we remove support for the prev env var? So this is a breaking change?

levkohimins replied:
No. I've added backwards compatibility for these env vars, and explained why I did it [here](https://github.com/gruntwork-io/terragrunt/pull/2608#issuecomment-1631348902)

https://github.com/gruntwork-io/terragrunt/blob/7f242e9ffbfe8b8f773fa8fe59d7303e8c9f8793/cli/app.go#L115-L118

brikis98 commented on docs/_docs/02_features/auto-retry.md:
````diff
@@ -52,4 +52,4 @@ retry_max_attempts = 5
 ```
 
-To disable `auto-retry`, use the `--terragrunt-no-auto-retry` command line option or set the `TERRAGRUNT_AUTO_RETRY` environment variable to `false`.
+To disable `auto-retry`, use the `--terragrunt-no-auto-retry` command line option or set the `TERRAGRUNT_NO_AUTO_RETRY` environment variable to `true`.
````
Same question here

brikis98 commented on docs/_docs/04_reference/cli-options.md:
```diff
@@ -543,7 +544,7 @@ disabled. See [Auto-Init]({{site.baseurl}}/docs/features/auto-init#auto-init)
 
 **CLI Arg**: `--terragrunt-no-auto-approve`<br/>
-**Environment Variable**: `TERRAGRUNT_AUTO_APPROVE` (set to `false`)
+**Environment Variable**: `TERRAGRUNT_NO_AUTO_APPROVE` (set to `true`)
```
Another breaking change?

levkohimins replied:
Explained above

brikis98 commented on docs/_docs/04_reference/cli-options.md:
```diff
@@ -564,7 +565,7 @@ When passed in, don't automatically retry commands which fail with transient err
 
 **CLI Arg**: `--terragrunt-non-interactive`<br/>
-**Environment Variable**: `TF_INPUT` (set to `false`)
+**Environment Variable**: `TERRAGRUNT_NON_INTERACTIVE` (set to `true`)
```
Was this a fix for a copy/paste error? Or was this renamed?

levkohimins replied:
At first I thought it was a copy/paste error (not mine), but now I realized this is Terraform env var [TF_INPUT](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_input)

As I just understood, as it was originally intended, if we disable terraform input by this env var, then this should also enable non-interactive mode in Terragrant
![Code search results · GitHub 2023-07-12 at 3 22 39 PM](https://github.com/gruntwork-io/terragrunt/assets/62853952/7f4327cf-db2f-46a2-bd6b-15c0db22f2af)

If yes, then it seems to me that it looks a bit unusual, doesn't it? Anyway, should I return this logic?

brikis98 replied:
Ah, good find. We should maintain backward compatibility with the `TF_INPUT` env var. We could additionally add a TG-specific env var for this as you have, but I wouldn't remove the `TF_INPUT` behavior or the mention in the docs. If anything, might be worth adding in the docs that `TF_INPUT` is native to TF!

levkohimins replied:
Agree! We should be ensure, that these improvements will not harm the workflow of customers who were using an older version. Kept backwards compatibility. Updated the docs.

brikis98 commented on options/options.go:
```diff
@@ -329,64 +347,75 @@ func NewTerragruntOptionsForTest(terragruntConfigPath string) (*TerragruntOption
+		FetchDependencyOutputFromState: opts.FetchDependencyOutputFromState,
+		UsePartialParseConfigCache:     opts.UsePartialParseConfigCache,
+		OutputPrefix:                   opts.OutputPrefix,
+		IncludeModulePrefix:            opts.IncludeModulePrefix,
```
NIT: this is a already a massive refactor (8K lines of code change!), and renames like this add a lot of lines of code for something that is largely unrelated. In the future, please do this sort of change (variable rename) in a separate PR from, say, refactoring how CLI commands/args are handled.

levkohimins replied:
Fair point, next time I will not clog the pull request with such changes. Sorry.

brikis98 commented on pkg/cli/app.go:
```diff
@@ -0,0 +1,89 @@
+package cli
```
Why does some stuff go into `pkg/cli` and other stuff in `cli`? How are you thinking about the `pkg` folder?

levkohimins replied:
> Are we using it to contain wrappers of other (e.g., open source) packages? If so, is `pkg` a Go idiom for that or something we came up with here?

Right. I should have explained it.  Added `/pkg/pkg.go` file with comment:

```go
// The purpose of this package `/pkg` is a developing a new common libraries which can be moved to `go-commons` in the future.
// This approach helps us test new functionality well on a single application first and after that to safely merged with rest common libraries in `go-commons`
```

brikis98 commented on pkg/cli/app.go:
```diff
@@ -0,0 +1,89 @@
+	"github.com/urfave/cli/v2"
+)
+
+// App is the main structure of app cli application. It should be created with the cli.NewApp() function.
+type App struct {
```
We should explain that it's a wrapper for `urfave`'s `cli.App` struct, and explain what we're adding here that's different.

levkohimins replied:
Fair point, explained.

brikis98 commented on pkg/cli/args.go:
```diff
@@ -0,0 +1,67 @@
+
+type NormalizeActsType byte
+
+type Args struct {
```
Same here... Explain we're wrapping the `urfave` package of the same name, and what we're adding.

brikis98 replied:
Looks like we wrap a bunch of other packages below too... Same comment applies to them.

levkohimins replied:
Agree, added unit tests and comments.

brikis98 commented on pkg/cli/args.go:
```diff
@@ -0,0 +1,67 @@
+	var strArgs []string
+
+	for _, arg := range args.Slice() {
+		arg := arg
```
Why do we need this? I don't see any goroutine or other closure that requires us to rescope this?

levkohimins replied:
Ah, I just forgot to remove, after modifying the code. Thanks!

brikis98 commented on pkg/cli/args.go:
```diff
@@ -0,0 +1,67 @@
+					arg = arg[1:]
+				}
+			case DoubleDashFlag:
+				if len(arg) >= 2 && arg[0] == '-' && arg[1] != '-' {
```
`-[^-]`

levkohimins replied:
Thanks for the suggestion, agree, changed.

brikis98 commented on pkg/env/env.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+// GetIntEnv returns the environment value converted to boolean type, or returns the specified fallback value if the variable with the given key is not present.
+func GetBoolEnv(key string, fallback bool) bool {
```
These seem like they might be better off in go-commons. Not urgent...

levkohimins replied:
Yes, that's how it was intended. I will move this code after merging the PR. 
_(Adding directly to go-commons at the very beginning is not always convenient for work, because sometimes it needs to be changed quickly)_

brikis98 commented on pkg/env/env_test.go:
```diff
@@ -0,0 +1,179 @@
+				runTestCase()
+			}
+
+			os.Setenv(envVarName, testCase.envVarValue)
```
I would strongly recommend against setting env vars in test cases. The env vars are global, but test cases often have concurrency, so this will lead to sporadic test behavior depending on which tests happen to run concurrently with what other tests.

I'd instead update the env package to work with the env vars passed in, rather than read directly from `os.Getenv`. Tests can then pass in mock values; real code can pass in real values.

levkohimins replied:
Thank you! Indeed your approach is better. Redone.

brikis98 commented on pkg/maps/maps.go:
```diff
@@ -0,0 +1,24 @@
+	"strings"
+)
+
+func Join[M ~map[K]V, K comparable, V any](m M, sliceSep, mapSep string) string {
```
Another go-commons candidate.

Needs a comment explaining what it does.

levkohimins replied:
Added

brikis98 commented on pkg/maps/maps.go:
```diff
@@ -0,0 +1,24 @@
+	return strings.Join(list, sliceSep)
+}
+
+func Slice[M ~map[K]V, K comparable, V any](m M, sep string) []string {
```
Same here

levkohimins replied:
Added

brikis98 reviewed (commented):
Wow, what a huge PR! That... was hard to review. Not sure I could cover everything.

A few questions:

1. What has changed about logging? In particular, has anything changed about what gets written to `stdout` on any command? TG is often used in scripts and people parse `stdout` carefully.
2. What has changed about the CLI flags or env vars TG expects? I saw some env vars that looked like backward incompatible changes.
3. Has anything changed in CLI flag parsing? Are we parsing differently than before? E.g., Different flag order? Different usage of `-xxx` versus `--xxx`?

@robmorgan Does this affect any of the TG APIs we use in Patcher?

levkohimins commented on .circleci/config.yml:
```diff
@@ -68,6 +68,7 @@ jobs:
       - store_test_results:
           path: logs
   integration_test:
+    resource_class: large
```
This change was made in this [PR](https://github.com/gruntwork-io/terragrunt/pull/2609)
but i ran into some issue with CircleCi and revoked this change to figure out the issue and then re-added. That's why it looks like my change.

levkohimins commented:
Thank you for the review and sorry for such the huge PR!

> 1. What has changed about logging? In particular, has anything changed about what gets written to `stdout` on any command? TG is often used in scripts and people parse `stdout` carefully.

Only falg `--terragrunt-no-color` has been added to disable terragrunt colors (to solve this issue https://github.com/gruntwork-io/terragrunt/issues/2116
any), no other changes affecting log output have been made.

> 2. What has changed about the CLI flags or env vars TG expects? I saw some env vars that looked like backward incompatible changes.

There were three flags that disabled some features, but their env vars behavior was the opposite:

```
--terragrunt-no-auto-init TERRAGRUNT_AUTO_INIT
--terragrunt-no-auto-retry TERRAGRUNT_AUTO_RETRY
--terragrunt-no-auto-approve TERRAGRUNT_AUTO_APPROVE
```
and these env vars only work with `false` value, for example `TERRAGRUNT_AUTO_INIT=true` does do nothing.

I renamed env vars and made their behavior more logical:
```
--terragrunt-no-auto-init TERRAGRUNT_NO_AUTO_INIT
--terragrunt-no-auto-retry TERRAGRUNT_NO_AUTO_RETRY
--terragrunt-no-auto-approve TERRAGRUNT_NO_AUTO_APPROVE
```

at the same time I left these old env variables for backwards compatibility. For example, both these env variables:

`TERRAGRUNT_AUTO_INIT=false`
`TERRAGRUNT_NO_AUTO_INIT=true`

are available and their behavior is identical. By the way, now all env variables are also displayed in the help https://share.getcloudapp.com/NQupYyzz


> 3. Has anything changed in CLI flag parsing? Are we parsing differently than before? E.g., Different flag order? Different usage of `-xxx` versus `--xxx`?

Yes, and for the better. Now it doesn't matter to use one hyphen or two. The principle is as follows, if the flag is not registered in terragrant, then it is considered as a terraform flag.
Of cource full backward compatibility. I used the `flag` package for parsing https://pkg.go.dev/flag#hdr-Command_line_flag_syntax

levkohimins commented on cli/app.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// This set of commands is also used in unit tests
+func newCommands(opts *options.TerragruntOptions) cli.Commands {
```
Agree with your remark, indeed, `terragruntCommands` says much more about what the func returns. Renamed.

levkohimins commented on cli/app.go:
```diff
@@ -0,0 +1,215 @@
+		return cli.ShowCommandHelp(ctx, ctx.Command.Name)
+	}
+
+	// If the first argument is a command, it is most likely a terraform command, show Terraform help.
+	if cmdName := ctx.Args().CommandName(); cmdName != "" {
+		terraformHelpCmd := append([]string{cmdName, "-help"}, ctx.Args().Tail()...)
+		return shell.RunTerraformCommand(opts, terraformHelpCmd...)
+	}
```
We had the test for `--help`, but after your comment, I realized that I forgot to add the `--version` flag.
Added flag and test for it. Thanks!

levkohimins commented on pkg/cli/args.go:
```diff
@@ -0,0 +1,67 @@
+		for _, act := range acts {
+			switch act {
+			case OneDashFlag:
+				if len(arg) >= 3 && arg[0:2] == "--" && arg[2] != '-' {
```
Thanks for the suggestion, agree, changed.

brikis98 reviewed (dismissed):
Thanks for this clean up. This looks like a solid improvement.

Given the size/scope of the PR, I'd be grateful for one more set of eyes before merge. @denis256 Could you take a look and share your thoughts?

brikis98 commented:
(Oh, and there is some test failure? Make sure test pass before merge.)

denis256 commented:
Noticed that build version is reported differently when terragrunt is built locally:
master branch
```
$ make terragrunt
set -xe ;\
vtag_maybe_extra=$(git describe --tags --abbrev=12 --dirty --broken) ;\
go build -o terragrunt -ldflags "-X main.VERSION=${vtag_maybe_extra}" .
++ git describe --tags --abbrev=12 --dirty --broken
+ vtag_maybe_extra=v0.48.6
+ go build -o terragrunt -ldflags '-X main.VERSION=v0.48.6' .

$ ./terragrunt --version
terragrunt version v0.48.6

```

feature/refactoring-cli-related-code branch
```
$ make terragrunt
set -xe ;\
vtag_maybe_extra=$(git describe --tags --abbrev=12 --dirty --broken) ;\
go build -o terragrunt -ldflags "-X main.VERSION=${vtag_maybe_extra}" .
++ git describe --tags --abbrev=12 --dirty --broken
+ vtag_maybe_extra=v0.48.5-94-g04ddc2eea20e
+ go build -o terragrunt -ldflags '-X main.VERSION=v0.48.5-94-g04ddc2eea20e' .

$ ./terragrunt --version
terragrunt version latest

```

denis256 commented:
master
```
$ terragrunt
<print help>
```

feature/refactoring-cli-related-code branch
```
$ terragrunt
ERRO[0000] Missing terraform command (Example: terragrunt plan) 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
```

levkohimins commented:
> master
> 
> ```
> $ terragrunt
> <print help>
> ```
> 
> feature/refactoring-cli-related-code branch
> 
> ```
> $ terragrunt
> ERRO[0000] Missing terraform command (Example: terragrunt plan) 
> ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
> ```

Fixed!

> Noticed that build version is reported differently when terragrunt is built locally: master branch
> 
> ```
> $ make terragrunt
> set -xe ;\
> vtag_maybe_extra=$(git describe --tags --abbrev=12 --dirty --broken) ;\
> go build -o terragrunt -ldflags "-X main.VERSION=${vtag_maybe_extra}" .
> ++ git describe --tags --abbrev=12 --dirty --broken
> + vtag_maybe_extra=v0.48.6
> + go build -o terragrunt -ldflags '-X main.VERSION=v0.48.6' .
> 
> $ ./terragrunt --version
> terragrunt version v0.48.6
> ```
> 
> feature/refactoring-cli-related-code branch
> 
> ```
> $ make terragrunt
> set -xe ;\
> vtag_maybe_extra=$(git describe --tags --abbrev=12 --dirty --broken) ;\
> go build -o terragrunt -ldflags "-X main.VERSION=${vtag_maybe_extra}" .
> ++ git describe --tags --abbrev=12 --dirty --broken
> + vtag_maybe_extra=v0.48.5-94-g04ddc2eea20e
> + go build -o terragrunt -ldflags '-X main.VERSION=v0.48.5-94-g04ddc2eea20e' .
> 
> $ ./terragrunt --version
> terragrunt version latest
> ```

This was also changed, please take a look at `.circleci/config.yml` file.  The new implementation looks like this 
```shell
... -ldflags="-X github.com/gruntwork-io/go-commons/version.Version=${vtag_maybe_extra} -extldflags '-static'"
```

This is a more advanced implementation, also used in `repo-copier` and `terraform-aws-lambda`.

levkohimins commented:
> master
> 
> ```
> $ terragrunt
> <print help>
> ```
> 
> feature/refactoring-cli-related-code branch
> 
> ```
> $ terragrunt
> ERRO[0000] Missing terraform command (Example: terragrunt plan) 
> ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
> ```

Fixed. Thanks.

levkohimins commented:
> Thanks for this clean up. This looks like a solid improvement.
> 
> Given the size/scope of the PR, I'd be grateful for one more set of eyes before merge. @denis256 Could you take a look and share your thoughts?

Thank you for your review! You helped me improve the code where I was lazy :) 

> (Oh, and there is some test failure? Make sure test pass before merge.)

Of course! BTW I have fixed one very unpleasant bug in the integration tests. Before the refactoring, it appeared very rarely, and just restarting the CircleCi task was enough, but after that, this bug became permanent. (@denis256 FYI)

https://github.com/gruntwork-io/terragrunt/blob/714dcc0dc6536013f56f0c76585b5d004e15a6bd/test/integration_test.go#L4648-L4650

levkohimins commented on cli/app_test.go:
```diff
@@ -0,0 +1,461 @@
+}
+
+func TestTerraformHelp_wrongHelpFlag(t *testing.T) {
+	app := NewApp(os.Stdout, os.Stderr)
```
Fixed

levkohimins commented on cli/app_test.go:
```diff
@@ -0,0 +1,461 @@
+	return fmt.Sprintf("--%s", name)
+}
+
+type argMissingValue string
```
Fixed

levkohimins commented:
@MoonMoon1919 Thank you for the review!

denis256 commented:
Noticed another difference:
https://github.com/denis256/terragrunt-tests/tree/master/output-module-groups/app

```
$ master-terragrunt output-module-groups
Module:                  /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/app
External dependency:     /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4
Should Terragrunt apply the external dependency? (y/n) 

$ branch-terragrunt output-module-groups
Module /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/app depends on module /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4, which is an external dependency outside of the current working directory. Should Terragrunt run this external dependency? Warning, if you say 'yes', Terragrunt will make changes in /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4 as well! (y/n) 

```

levkohimins commented:
> Noticed another difference: https://github.com/denis256/terragrunt-tests/tree/master/output-module-groups/app
> 
> ```
> $ master-terragrunt output-module-groups
> Module:                  /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/app
> External dependency:     /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4
> Should Terragrunt apply the external dependency? (y/n) 
> 
> $ branch-terragrunt output-module-groups
> Module /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/app depends on module /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4, which is an external dependency outside of the current working directory. Should Terragrunt run this external dependency? Warning, if you say 'yes', Terragrunt will make changes in /home/denis/projects/gruntwork/terragrunt-tests/output-module-groups/m4 as well! (y/n) 
> ```

Merged with `master`. Now should be the same.

denis256 commented on pkg/cli/args.go:
```diff
@@ -0,0 +1,78 @@
+	// This is the only way to avoid duplicating code from the private struct `urfave`'s `cli.args` that implements `cli.Args` interface.
+	flagSet := libflag.NewFlagSet("", libflag.ContinueOnError)
+	flagSet.SetOutput(io.Discard)
+	flagSet.Parse(append([]string{"--"}, args...))
```
Error returned from `Parse()` seems to be unhandled...
```
func (f *FlagSet) Parse(arguments []string) error
```

levkohimins replied:
Redone to get rid of this feature altogether

levkohimins commented on cli/commands/terraform/action.go:
```diff
@@ -0,0 +1,714 @@
+		for _, arg_cmd := range arg.Commands {
+			if cmd == arg_cmd {
+				lastArg := util.LastArg(terragruntOptions.TerraformCliArgs)
+				skipVars := (cmd == "apply" || cmd == CommandNameDestroy) && util.IsFile(lastArg)
```
Replaced

levkohimins commented on pkg/cli/flag.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+var (
+	// use to separate arguments and env vars with multiple values.
```
Reworded

levkohimins commented on pkg/cli/flag.go:
```diff
@@ -0,0 +1,104 @@
+	return ""
+}
+
+func value(envKey string) *string {
```
Removed

levkohimins commented on pkg/cli/map_flag.go:
```diff
@@ -0,0 +1,214 @@
+
+var (
+	// use to separate arguments and env vars with multiple values.
+	DefaultSplitter = strings.Split
```
Removed

levkohimins commented on pkg/cli/map_flag.go:
```diff
@@ -0,0 +1,214 @@
+	return []string{flag.EnvVar}
+}
+
+// GetValue returns the flags value as string representation and an empty string if the flag takes no value at all.
```
Renamed

levkohimins commented on pkg/env/env_test.go:
```diff
@@ -0,0 +1,183 @@
+	}
+}
+
+func TestParseEnvironmentVariables(t *testing.T) {
```
Added
